### PR TITLE
CVS-77193 DAG layout intersections & binary image handling changes

### PR DESCRIPTION
**Pull request by dkalinowski**

### CODE REVIEW

**Comment by dtrawins** [comment]
We should prevent binary image resize when the model layout is different from N..HWC. That will prevent unexpected interpretation of the dimensions in the resize.

File: src/binaryutils.cpp [hunk 1]
```diff
@@ -89,79 +89,25 @@ Status convertPrecision(const cv::Mat& src, cv::Mat& dst, const ovms::Precision
 }
 
 Status validateLayout(const std::shared_ptr<TensorInfo>& tensorInfo) {
-    if ((tensorInfo->getLayout() != "NHWC") &&
-        (tensorInfo->getLayout() != Layout::getUnspecifiedLayout()) &&  // handle DAG
-        (tensorInfo->getLayout() != Layout::getDefaultLayout())) {      // handle model without Layout set
+    static const std::string binarySupportedLayout = "N...HWC";
+    if (!tensorInfo->getLayout().createIntersection(Layout(binarySupportedLayout), tensorInfo->getShape().size()).has_value()) {
+        SPDLOG_ERROR("Endpoint needs to be compatible with {} to support binary image inputs, actual: {}",
```
Comment (atobiszei) [inline src/binaryutils.cpp:94]: Shouldn;t this be DEBUG? We could spam OVMS logs with predict requests otherwise

File: src/binaryutils.cpp [hunk 2]
```diff
@@ -258,12 +186,64 @@ Status validateTensor(const std::shared_ptr<TensorInfo>& tensorInfo,
...
 Status convertTensorToMatsMatchingTensorInfo(const tensorflow::TensorProto& src, std::vector<cv::Mat>& images, const std::shared_ptr<TensorInfo>& tensorInfo) {
+    Dimension tensorInfoHeightDim = getTensorInfoHeightDim(tensorInfo);
+    Dimension tensorInfoWidthDim = getTensorInfoWidthDim(tensorInfo);
+
     for (int i = 0; i < src.string_val_size(); i++) {
         cv::Mat image = convertStringValToMat(src.string_val(i));
         if (image.data == nullptr)
             return StatusCode::IMAGE_PARSING_FAILED;
 
+        updateAlignmentResolution(tensorInfoHeightDim, tensorInfoWidthDim, image);
```
Comment (atobiszei) [inline src/binaryutils.cpp:245]: getAlignedResolution?

File: src/pipelinedefinition.cpp [hunk 3]
```diff
@@ -1065,14 +1065,17 @@ std::shared_ptr<TensorInfo> createOutputTensorInfoForPipeline(const std::string&
...
-    auto newTensorInfo = std::make_shared<TensorInfo>(tensorInfo);
+    auto newTensorInfo = std::make_shared<TensorInfo>(alias, tensorInfo.getPrecision(), tensorInfo.getShape(), tensorInfo.getLayout());
     auto it = inputsInfo.find(alias);
     if (it != inputsInfo.end()) {
         if (!it->second->isTensorSpecEqual(*newTensorInfo)) {
             auto intersectionTensorInfo = it->second->createIntersection(*newTensorInfo);
             if (intersectionTensorInfo == nullptr) {
                 Status status = StatusCode::PIPELINE_INPUTS_AMBIGUOUS_METADATA;
-                SPDLOG_LOGGER_ERROR(modelmanager_logger, "Error validating pipeline: {}", status.string());
+                SPDLOG_LOGGER_ERROR(modelmanager_logger, "Error validating pipeline: {}; {}; {}",
```
Comment (atobiszei) [inline src/pipelinedefinition.cpp:1075]: More explanation within error message

File: src/binaryutils.cpp (hunk 1 above)
Comment (dkalinowski) [reply src/binaryutils.cpp:94]: There are plenty of non-debug logs when request is compatible with endpoint.
Examples when precision or shape is wrong:

```
[2022-02-14 12:34:50.702][832][serving][warning][modelinstance.cpp:957] Model: resnet, version: 1 validation of inferRequest failed. Status Code: 61, Error: Invalid input precision - Expected: FP32; Actual: FP64; input name: 0
[2022-02-14 12:35:09.896][833][serving][warning][modelinstance.cpp:957] Model: resnet, version: 1 validation of inferRequest failed. Status Code: 60, Error: Invalid input shape - Expected: (1,3,224,224); Actual: (1,4,224,224); input name: 0
```

If we want to move all errors to DEBUG, this should be separate task. @atobiszei

File: src/binaryutils.cpp (hunk 2 above)
Comment (dkalinowski) [reply src/binaryutils.cpp:245]: This updates the resolution, therefore `getAlignedResolution` would be misleading

File: src/pipelinedefinition.cpp (hunk 3 above)
Comment (dkalinowski) [reply src/pipelinedefinition.cpp:1075]: The error is actually explained in depth, because we print `status.string()`.

File: src/binaryutils.cpp (hunk 2 above)
Comment (dkalinowski) [reply src/binaryutils.cpp:245]: `updateTargetResolution`

File: src/pipelinedefinition.cpp (hunk 3 above)
Comment (dkalinowski) [reply src/pipelinedefinition.cpp:1075]: TODO: Add new line between tensors

File: src/binaryutils.cpp (hunk 2 above)
Comment (dkalinowski) [reply src/binaryutils.cpp]: Done

File: src/pipelinedefinition.cpp (hunk 3 above)
Comment (dkalinowski) [reply src/pipelinedefinition.cpp]: Added

File: src/test/ensemble_metadata_test.cpp [hunk 4]
```diff
@@ -316,6 +316,131 @@ TEST(EnsembleMetadata, TwoParallelModelsOneShapeAnyOneStaticIntersection) {
...
+    ASSERT_NE(outputs.find("request_output_name_B"), outputs.end());
+
+    const auto& input = inputs.at("request_input_name");
+    EXPECT_EQ(input->getLayout(), ovms::Layout{"N?C"});
+
+    const auto& output_A = outputs.at("request_output_name_A");
+    EXPECT_EQ(output_A->getLayout(), ovms::Layout{"..."});
+
+    const auto& output_B = outputs.at("request_output_name_B");
+    EXPECT_EQ(output_B->getLayout(), ovms::Layout{"..."});
```
Comment (atobiszei) [inline src/test/ensemble_metadata_test.cpp:441]: Shouldn;t we have here
N...
N...C
or at least
...
...C?

File: src/test/ensemble_tests.cpp [hunk 5]
```diff
@@ -5329,7 +5329,7 @@ static const char* pipelineSingleIncrement4DimOutputNHWCRangeResolutionEntryDemu
         {
             "name": "increment_pipeline",
             "inputs": ["pipeline_input"],
-            "demultiply_count": 0,
```
Comment (atobiszei) [inline src/test/ensemble_tests.cpp:5332]: Why this change?

File: src/test/layout_test.cpp [hunk 6]
```diff
@@ -64,26 +64,20 @@ TEST(Layout, Validate) {
...
-    EXPECT_EQ(Layout("N...").createIntersection(Layout("??D")), Layout("N?D"));
+    EXPECT_EQ(Layout("NCHW").createIntersection(Layout("NCHW"), 4), Layout("NCHW"));
+    EXPECT_EQ(Layout("NCHWD").createIntersection(Layout("NCHW"), 5), std::nullopt);
+    EXPECT_EQ(Layout("NCHW").createIntersection(Layout("NCHWD"), 5), std::nullopt);
+    EXPECT_EQ(Layout("NCHW").createIntersection(Layout("N..."), 4), Layout("NCHW"));
+    EXPECT_EQ(Layout("N...").createIntersection(Layout("NCHW"), 4), Layout("NCHW"));
+    EXPECT_EQ(Layout("NCHWD").createIntersection(Layout("NCHW?"), 5), Layout("NCHWD"));
+    EXPECT_EQ(Layout("NCHW?").createIntersection(Layout("NCHWD"), 5), Layout("NCHWD"));  // test symmetry
+    EXPECT_EQ(Layout("NCHWD").createIntersection(Layout("NCHW"), 5), std::nullopt);
+    EXPECT_EQ(Layout("NCHW").createIntersection(Layout("NCHWD"), 5), std::nullopt);
```
Comment (atobiszei) [inline src/test/layout_test.cpp:75]: All negative cases have different number of dimensions than expected.

Please add at least one case with non matching layouts but all defined,
one with "...", and one with ?

File: src/test/ensemble_metadata_test.cpp (hunk 4 above)
Comment (dkalinowski) [reply src/test/ensemble_metadata_test.cpp:441]: N... => N?...
NC => N?C

Comment (dkalinowski) [reply src/test/ensemble_metadata_test.cpp:441]: TODO: Add task

File: docs/binary_input.md [hunk 7]
```diff
@@ -19,11 +19,15 @@ is out of range it will be adjusted to the nearer border. For example, when mode
 - if input shape is [1,90,200,3] it will be resized into [1,100,200,3]
 - if input shape is [1,220,200,3] it will be resized into [1,200,200,3]
 
-Processing the binary image requests requires the model or the custom nodes to accept NHWC layout in BGR color 
+When the endpoint (model or pipeline) shape supports any dimension value (`-1`), the image is not resized and it is enforced to have resolution aligned for entire batch of images.
```
Comment (dtrawins) [inline docs/binary_input.md:22]:
```suggestion
When the endpoint (model or pipeline) shape includes a dynamic dimension value (`-1`), the image is not resized and it is enforced to have identical resolution for a whole batch of images.
```

File: docs/binary_input.md [hunk 8]
```diff
@@ -19,11 +19,15 @@ is out of range it will be adjusted to the nearer border. For example, when mode
...
+
+Resize functionality is also disabled by default for endpoints with layouts different than `...` and `NHWC` (or `N?HWC`, when modified by demultiplexer).
```
Comment (dtrawins) [inline docs/binary_input.md:24]:
```suggestion
Resize functionality is also disabled for endpoints with layouts different than `...` and `NHWC` (or `N?HWC`, when modified by a [demultiplexer](demultiplexing.md) ).
```

File: src/test/binaryutils_test.cpp [hunk 9]
```diff
@@ -338,18 +346,52 @@ TEST_F(BinaryUtilsTest, positive_resizing_with_any_shape) {
...
     size_t expectedRowsNumber = 1;
     EXPECT_EQ(tensorDims[1], expectedRowsNumber);
-    size_t expectedColsNumber = 4;
+    size_t expectedColsNumber = 1;
     EXPECT_EQ(tensorDims[2], expectedColsNumber);
-    ASSERT_EQ(tensor.get_size(), 12);
+    ASSERT_EQ(tensor.get_size(), 3);
+}
+
+TEST_F(BinaryUtilsTest, positive_resizing_with_demultiplexer_and_range_resolution) {
```
[the remark this answers is not in the document]
Comment (dkalinowski) [reply src/test/binaryutils_test.cpp:371]: TODO: Add test to have image between min and max

File: docs/binary_input.md (hunk 8 above)
Comment (dtrawins) [inline docs/binary_input.md:24]: With such layout, I guess the binary format is not supported at all

File: src/test/layout_test.cpp (hunk 6 above)
Comment (dkalinowski) [reply src/test/layout_test.cpp:75]: TODO: have more cases

File: docs/binary_input.md [hunk 10]
```diff
@@ -19,11 +19,15 @@ is out of range it will be adjusted to the nearer border. For example, when mode
...
+
+Processing the binary image requests requires the model or the custom nodes have layout compatible with `N...HWC` and BGR color
```
Comment (dtrawins) [inline docs/binary_input.md:26]:
```suggestion
Requests with binary images can be processed for models or the custom nodes with the input layout `NHWC` or `N...HWC`. The models' inputs should use BGR color format.
```

File: docs/binary_input.md [hunk 11]
```diff
@@ -19,11 +19,15 @@ is out of range it will be adjusted to the nearer border. For example, when mode
...
+When the endpoint (model or pipeline) shape supports any dimension value (`-1`), the image is not resized and it is enforced to have resolution aligned for entire batch of images.
+
+Resize functionality is also disabled by default for endpoints with layouts different than `...` and `NHWC` (or `N?HWC`, when modified by demultiplexer).
+
+Processing the binary image requests requires the model or the custom nodes have layout compatible with `N...HWC` and BGR color 
 format with data with the data range from 0-255. Original layout of the input data can be changed in the 
 OVMS configuration in runtime. For example when the orignal model has input shape [1,3,224,224] add a parameter
 in the OVMS configuration "layout": "NHWC:NCHW" or the command line parameter `--layout NHWC:NCHW`. In result, the model will
-have effective shape [1,224,224,3].
+have effective shape [1,224,224,3] and layout `NHWC`, which is compatible with `N...HWC`.
```
Comment (dtrawins) [inline docs/binary_input.md:30]:
```suggestion
have effective shape [1,224,224,3] and layout `NHWC`.
```

File: docs/binary_input.md (hunk 8 above)
Comment (dkalinowski) [reply docs/binary_input.md:24]: Binary format is supported for those layouts. NHWC, N?NHWC (in case of demultiplexer) and ... (in case of single custom node at the DAG entry)

Comment (dkalinowski) [reply docs/binary_input.md:24]: What do you mean by that?

File: docs/binary_input.md (hunk 10 above)
Comment (dkalinowski) [reply docs/binary_input.md:26]: This is not true, requests with binary images can be processed for any layout that is compatible with `N...HWC`, as stated in my change.
This means, binary images are supported for any endpoint with layout that can be intersected with `N...HWC`, for example:
- ... (custom node)
- N... (default model layout)
- N?... (default model layout + demultiplexer)
- NH... (--layout NH...)
- NHW... (--layout NHW...)
- NHWC  (--layout NHWC)
- N?HWC (--layout NHWC + demultiplexer)

**Review by atobiszei — APPROVED** [review]

**Review by dtrawins — APPROVED** [review]

File: docs/binary_input.md (hunk 11 above)
Comment (dkalinowski) [reply docs/binary_input.md]: done

File: docs/binary_input.md (hunk 10 above)
Comment (dkalinowski) [reply docs/binary_input.md]: Improved the doc
